### PR TITLE
fix(pu): fix recompute advantage in on policy ppo and polish rnd_onppo algorithm

### PR DESCRIPTION
**Pull request by puyuan1996**
## Description
fix recompute advantage in on policy ppo, add onppo config for minigrid and mujoco env,
and polish rnd_onppo algorithm.

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: dizoo/box2d/lunarlander/config/lunarlander_rnd_onppo_config.py [hunk 1]
```diff
@@ -1,39 +1,54 @@
...
+            # nstep_return=False,
             adv_norm=True,
+            value_norm=True,
         ),
         collect=dict(
-            n_sample=128,
+            # n_sample=128,
+            collector_env_num=collector_env_num,
+            n_sample=int(64 * collector_env_num),
+            #  self._traj_len  = max(1,64*8//8)=64
```
Comment (PaParaZz1) [inline dizoo/box2d/lunarlander/config/lunarlander_rnd_onppo_config.py:47]: remove unused commented config item

File: ding/utils/default_helper.py [hunk 2]
```diff
@@ -441,8 +442,16 @@ def split_data_generator(data: dict, split_size: int, shuffle: bool = True) -> d
                 batch[k] = None
             elif k.startswith('prev_state'):
                 batch[k] = [data[k][t] for t in indices[i:i + split_size]]
-            elif isinstance(data[k], list) or isinstance(data[k], tuple):
-                batch[k] = [t[indices[i:i + split_size]] for t in data[k]]
+            # elif isinstance(data[k], list) or isinstance(data[k], tuple):
+            #     batch[k] = [t[indices[i:i + split_size]] for t in data[k]]
+            elif k == 'logit':  # for continuous action
```
Comment (PaParaZz1) [inline ding/utils/default_helper.py:447]: keep compatible, add new if-else branch in `isinstance(data[k], list) or isinstance(data[k], tuple)`

File: ding/reward_model/rnd_reward_model.py [hunk 3]
```diff
@@ -69,10 +71,21 @@ def __init__(self, config: EasyDict, device: str, tb_logger: 'SummaryWriter') ->
...
     def _train(self) -> None:
         train_data: list = random.sample(self.train_data, self.cfg.batch_size)
         train_data: torch.Tensor = torch.stack(train_data).to(self.device)
+
+        # TODO(pu): observation normalization:  transform to mean 0, std 1
+        self._running_mean_std_rnd_obs.update(train_data.cpu().numpy())
+        train_data = (train_data - to_tensor(self._running_mean_std_rnd_obs.mean).to(self.device)) / to_tensor(
+            self._running_mean_std_rnd_obs.std
+        ).to(self.device)
+        train_data = torch.clamp(train_data, min=-5, max=5)
```
Comment (PaParaZz1) [inline ding/reward_model/rnd_reward_model.py:87]: add clamp min and max in rnd default_config

File: ding/reward_model/rnd_reward_model.py [hunk 4]
```diff
@@ -69,10 +71,21 @@ def __init__(self, config: EasyDict, device: str, tb_logger: 'SummaryWriter') ->
...
+        self.estimate_cnt_rnd = 0
+        self._running_mean_std_rnd_obs = RunningMeanStd(epsilon=1e-4)  # TODO(pu)
 
     def _train(self) -> None:
         train_data: list = random.sample(self.train_data, self.cfg.batch_size)
         train_data: torch.Tensor = torch.stack(train_data).to(self.device)
+
+        # TODO(pu): observation normalization:  transform to mean 0, std 1
+        self._running_mean_std_rnd_obs.update(train_data.cpu().numpy())
+        train_data = (train_data - to_tensor(self._running_mean_std_rnd_obs.mean).to(self.device)) / to_tensor(
```
Comment (PaParaZz1) [inline ding/reward_model/rnd_reward_model.py:84]: add a obs_norm option

File: ding/reward_model/rnd_reward_model.py [hunk 5]
```diff
@@ -82,27 +95,52 @@ def _train(self) -> None:
     def train(self) -> None:
         for _ in range(self.cfg.update_per_collect):
             self._train()
-        self.clear_data()
+        # self.clear_data()
```
Comment (PaParaZz1) [inline ding/reward_model/rnd_reward_model.py:98]: why comment this line code

File: ding/reward_model/rnd_reward_model.py [hunk 6]
```diff
@@ -69,10 +71,21 @@ def __init__(self, config: EasyDict, device: str, tb_logger: 'SummaryWriter') ->
         assert self.intrinsic_reward_type in ['add', 'new', 'assign']
         self.train_data = []
         self.opt = optim.Adam(self.reward_model.predictor.parameters(), config.learning_rate)
+        self._running_mean_std_rnd = RunningMeanStd(epsilon=1e-4)
```
Comment (PaParaZz1) [inline ding/reward_model/rnd_reward_model.py:74]: rename to `self._running_mean_std_reward` and `self._running_mean_std_obs`

File: ding/reward_model/rnd_reward_model.py [hunk 7]
```diff
@@ -82,27 +95,52 @@ def _train(self) -> None:
...
+            self.tb_logger.add_scalar('rnd_reward/rnd_reward_max', rnd_reward.max(), self.estimate_cnt_rnd)
+            self.tb_logger.add_scalar('rnd_reward/rnd_reward_mean', rnd_reward.mean(), self.estimate_cnt_rnd)
+            self.tb_logger.add_scalar('rnd_reward/rnd_reward_min', rnd_reward.min(), self.estimate_cnt_rnd)
+
+            rnd_reward = rnd_reward.to(data[0]['reward'].device)
+            rnd_reward = torch.chunk(rnd_reward, rnd_reward.shape[0], dim=0)
+        for item, rnd_rew in zip(data, rnd_reward):
             if self.intrinsic_reward_type == 'add':
-                item['reward'] += rew
+                if item['reward'] > 0 and item['reward'] <= 1:
```
Comment (PaParaZz1) [inline ding/reward_model/rnd_reward_model.py:135]: add a `extrinsic_reward_scale` option, defaults to 1.0

File: ding/policy/ppo.py [hunk 8]
```diff
@@ -212,6 +375,7 @@ def _forward_learn(self, data: Dict[str, Any]) -> Dict[str, Any]:
 
                 self._optimizer.zero_grad()
                 total_loss.backward()
+                torch.nn.utils.clip_grad_norm_(self._model.parameters(), 10)  # TODO(pu)
```
Comment (PaParaZz1) [inline ding/policy/ppo.py:378]: add `grad_clip` and `clip_value` option, defaults to False and 10

File: ding/policy/ppo.py [hunk 9]
```diff
@@ -255,6 +419,7 @@ def _init_collect(self) -> None:
             Init traj and unroll length, collect model.
         """
         self._unroll_len = self._cfg.collect.unroll_len
+        self._unroll_len_add_burnin_step = self._unroll_len
```
Comment (PaParaZz1) [inline ding/policy/ppo.py:422]: remove burnin

File: ding/policy/ppo.py [hunk 10]
```diff
@@ -7,23 +7,148 @@
...
+def compute_adv(data, last_value, cfg):
+    data = get_gae(
+        data,
+        # to_device(last_value, 'cpu'),
+        last_value,
+        gamma=cfg.collect.discount_factor,
+        gae_lambda=cfg.collect.gae_lambda,
+        cuda=False
+    )  # data: list (T timestep, 1 batch) [['value':,'reward':,'adv':], ...,]
+    if not cfg.nstep_return:
```
Comment (PaParaZz1) [inline ding/policy/ppo.py:29]: use nstep_return as argument and replace if-else branch

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/reward_model/rnd_reward_model.py (hunk 5 above)
Comment (puyuan1996) [reply ding/reward_model/rnd_reward_model.py:98]: 在serial_pipeline_reward_model_onpolicy里面有设置
` if iter % cfg.reward_model.clear_buffer_per_iters == 0:
            reward_model.clear_data()`

**Review by PaParaZz1 — APPROVED** [review]
